kart/kartapi.py

```diff
@@ -229,11 +229,12 @@ def executeKart(commands, path=None, jsonoutput=False, feedback=None):
                     output.append(line)
                 stdout = "".join(output)
                 stderr = "".join(err)
+                proc.communicate()  # need to get the returncode
```

Comment:
mmm, this isn't right (neither was the existing code), and will end up deadlocking. It'll read everything from stderr until EOF; then everything from stdout until EOF; then (effectively) do a `p.wait()`. If the stdout OS buffer fills up before eof it'll hang.

If we need to read from both streams separately, we need to either use [selectors](https://stackoverflow.com/a/61585093/2662) which is simple but posix-only for pipes. Windows you need [threads](https://stackoverflow.com/a/4896288/2662) — that will work on posix too. Or the [asyncio way](https://stackoverflow.com/a/56051270/2662) runs cross-platform.

Or, since we're in QT — maybe much simpler to use [QProcess](https://doc.qt.io/qt-6/qprocess.html) which already deals with this afaics?

Reply:
I agree it's not pretty and these things should definitely be using the QProcess support - unfortunately a large clone that is successful locks up the UI at the moment as well.

There is also a gap in how errors are reported - not really in a way that is very user friendly. 

For the moment this change was just to get past the nasty failure case where the command has finished in a failure but that failure has no way of getting to the next layer as the returncode is never set, then the rest of a change on behaviours happens and so you end up with the plugin believing there is a repo which adds it to the repo tree which in turn is read each time something happens with a layer in a project (Kart layer or not) and throws errors.

Reply:
It would be nice to have a cleaner way of updating the progress bar, at the moment `_processProgressLine` does a bunch of inspection of the text coming back which could be fragile if anything else changes in Kart itself.

Reply:
https://docs.qgis.org/3.22/en/docs/pyqgis_developer_cookbook/tasks.html


That looks like the preferred way of QGIS doing things but has a big warning about not interacting with the UI from a task which might make the progress feedback tricky to do in a clean way.

Reply:
What's there works by coincidence.

Suggestion to make it clear it's broken and doesn't look like it works :-) 
```
        ) as proc:
            if feedback is not None:
                # FIXME: this is broken, if the process fills the stdout buffer it will hang here forever.
                # Needs to use QProcess or similar in a background task
                err = []
                for line in proc.stderr:
                    feedback(line)
                    err.append(line)
                stdout, _ = proc.communicate()
                stderr = "".join(err)
            else:
                stdout, stderr = proc.communicate()
```

Reply:
I'd be tempted to pass a timeout to .communicate() but potentially this operation will take a while.

ah... does progress go to stdout? Git outputting something like this will fill the buffer and hang:

Bytes read: 12345\r
Bytes read: 12346\r
Bytes read: 12347\r
Bytes read: 12348\r

Reply:
Good idea. I’ve opened an issue for a new approach and will put a link in the comments. There are a lot of corners where this will break. The happy path is ok-ish enough at the moment for another quick and dirty short term change but definitely should be short term.